For #508: correction of feed effect

paulodamaso opened the pull request:
For #508: correction of feed effect
- corrected feed effect to expire after next card evaluation
- corrected tests
- left puzzle to continue feed logic implementation in card play routine

paulodamaso commented:
@SilasReinagel This issue handles specifically about feed effects. I fixed all related to the effect, which means that it will expire when next card is played. The tests proves it, and the event for card evaluation was created too.

We are implementing the cards and their effects, but we don't have the card playing routine fully functional yet. That's why we haven't any working card in gameplay, albeit having a lot of cards already implemented. I'm just doing the same as I did with the other cards, I'm implementing the card and its effects having in mind the effect behaviors, not the card play behavior.

SilasReinagel commented:
@paulodamaso From ticket #508 `If the next card played by the Astromancer matches the Feed type, then the effect is resolved. Otherwise the effect expires and is wasted.`

The effect resolution isn't included in this piece of work, so it doesn't finish the feature required to close #508

If we need other tickets, feel free to create them. But as per the workflow policy, every ticket that is closed fully delivers the functionality specified in the ticket.

paulodamaso commented:
@SilasReinagel Rewrite the feed logic implementation:
- Created `SetFeedUpEffect` effect which sets up another effect in queue to be played if next card is for some feed type. All cards that have "Feed-X" effect must wrap the fed effect with this decorator passing `X` as parameter
- Created `TriggerFeedEffects` effect, which is responsible for triggering effects set up by `SetFeedUpEffect`. All cards that are supposed to take part on trigger logic must wrap its effects in an `TriggerFeedEffects` and pass the feed type that will execute the triggered behavior. This will make possible that every astromancer card that is played interacts with queued effects and trigger the one that have to be triggered for that moment
- Removed old `OnFeedEffect` effect and tests
- Fixed the cards that were using `OnFeedEffect` effects

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/EffectType.cs:
```diff
@@ -29,12 +29,14 @@ public enum EffectType
     CostPrimaryResourceEffect = 27,
     ForNumberOfTurns = 28,
-    FeedOnEffect = 29,
+    FeedOnEffect = 29, //unused effect
```
Go ahead and delete this if it's unused. As long as we don't reuse the number, and nothing is broken in game, we are good.

paulodamaso replied:
@SilasReinagel I'll comment all the line, for making sure that we won't reuse the number

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/OnNextCardEffect.cs:
```diff
@@ -0,0 +1,32 @@
+ * 
+ * Turn ending makes this effect active. 
+ */
+public class OnNextCardEffect : Effect
```
This isn't quite what is needed for Astromancer. 

We can about the Astromancer's next card, not any party next card.

paulodamaso replied:
@SilasReinagel I chose to make a generic effect, so we can reuse it later if needed.

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/SetFeedUpEffect.cs:
```diff
@@ -0,0 +1,47 @@
+    public void Apply(Member source, Target target)
+    {
+
+        Effect effect = new UnqueueAfterExceuteEffect(_effect);
```
This Queue pop mechanism seems, plus one-shot applying seems like a duplicated code idiom. 

Rather than repeat this, let's build it into a core object, and have the current Objects who use it delegate to the one-shot effect.

paulodamaso replied:
@SilasReinagel We already have this logic in `QueueEffect`. I'll use it.

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/SetFeedUpEffect.cs:
```diff
@@ -0,0 +1,47 @@
+        BattleEvent.Subscribe<FeedCardResolutionStarted>(
+            (msg) =>
+            {
+                if (Enum.GetName(typeof(FeedType), _feedType) == msg.CardFeedType)
```
You can just write `_feedType.ToString()`. It's shorter and easier to think about.

paulodamaso replied:
@SilasReinagel Fixed

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/SetFeedUpEffect.cs:
```diff
@@ -0,0 +1,47 @@
+        BattleEvent.Subscribe<FeedCardResolutionStarted>(
+            (msg) =>
+            {
+                if (Enum.GetName(typeof(FeedType), _feedType) == msg.CardFeedType)
```
Should not use `==` for string comparisons. Always use `.Equals` with case-sensitivity specification

paulodamaso replied:
@SilasReinagel Fixed

SilasReinagel commented on src/ProjectNeon/Assets/Features/Effects/SetFeedUpEffect.cs:
```diff
@@ -0,0 +1,44 @@
+    public void Apply(Member source, Target target)
+    {
+        new QueueEffect(_effect).Apply(source, target);
+        BattleEvent.Subscribe<FeedCardResolutionStarted>(
```
This one-shot idiom is still duplicated.

paulodamaso replied:
@SilasReinagel Duplicated how? I'ts the only occurence in the code

SilasReinagel replied:
Okay. I can refactor it later. Just put a TODO in here. No puzzle needed. It's not a whole unit of work.
